swiftsimio/initial_conditions/generate_particles.py

```diff
@@ -1,5 +1,1281 @@
+        othresh = erf(othresh)
+        if np.random.uniform() < othresh:
+
+            attempts = 0
+            while True:
+
+                attempts += 1
+                if attempts == nunder:
+                    break  # emergency halt
+
+                u = np.random.randint(0, nunder)
+                uind = indices[underdense][u]
+
+                if touched[uind]:
+                    continue  # skip touched particles
+
+                uthresh = (rhoA[uind] - rho[uind]) / rhoA[uind]
+
+                if np.random.uniform() < uthresh:
+                    # we have a match!
+                    # compute displacement for overdense particle
+                    dx = np.zeros(3, dtype=np.float)
+                    H = kernel_gamma * h[uind]
+                    for d in range(ndim):
+                        sign = 1 if np.random.random() < 0.5 else -1
+                        dx[d] = np.random.uniform() * 0.3 * H * sign
+
+                    x[oind] = x[uind] + dx
+                    touched[oind] = True
+                    touched[uind] = True
+                    moved += 1
+                    break
+
+    if moved > 0:
+
+        # check boundary conditions
+        if icSimParams["periodic"]:
+            for d in range(ndim):
+                x[x[:, d] > boxsize[d], d] -= boxsize[d]
+                x[x[:, d] < 0.0, d] += boxsize[d]
+        else:
+            for d in range(ndim):
+
+                # move them away from the edge by a random factor of mean "cell size" boxsize/npart^ndim
+                mask = x[:, d] > boxsize[d]
+                x[mask, d] = boxsize[d] * (
+                    1.0 - np.random.uniform(x[x > 1.0].shape) / npart ** (1.0 / ndim)
+                )
+
+                mask = x[:, d] < 0.0
+                x[mask, d] = (
+                    np.random.uniform(x[mask, d].shape)
+                    * boxsize[d]
+                    / npart ** (1.0 / ndim)
+                )
+
+    return x, indices[touched]
+
+
+def _IC_write_intermediate_output(
+    iteration: int,
+    x: np.ndarray,
+    m: np.ndarray,
+    rho: np.ndarray,
+    h: np.ndarray,
+    icSimParams: dict,
+):
+    r"""
+    Write an intermediate output of current particle positions, densities,
+    masses, and smoothing lengths.
+
+    Parameters
+    ----------------
+
+    iteration: int
+        current iteration number
+
+    x: np.ndarray
+        current particle coordinates
+
+    m: np.ndarray
+        particle masses
+
+    rho: np.ndarray
+        particle densities
+
+    h: np.ndarray
+        particle smoothing lengths
+
+    icSimParams: dict
+        a dict containing simulation parameters as returned by 
+        ``IC_set_IC_params()``.
+
+
+
+    Returns
+    -------------------
+
+    Nothing.
+
+    """
+
+    from swiftsimio import Writer
+
+    nx = icSimParams["nx"]
+    ndim = icSimParams["ndim"]
+    npart = nx ** ndim
+
+    ICunits = unyt.UnitSystem(
```

Comment:
I didn't think it was necessary. The only place where I demand units is in the ``boxsize`` (I added an explicit check now) parameter, the rest is voluntary. And if ``unit_l is None``, then take the same units as ``boxsize``.  The only reason I have a ``UnitSystem`` here is because from what I saw, ``Writer`` needs one.

Reply:
To make this consistent with the rest of `swiftsimio` it would be best to take a `UnitSystem` at the top level.

Reply:
It takes a ``unyt.unit_system.UnitSystem`` as the argument now.